rfc-146-production-deploy-access.md

```diff
@@ -0,0 +1,82 @@
+
+To avoid confusion, the existing "Production" access level should be renamed to "Production Admin" access.
+
+## Problem
+
+There is [a defined set of rules for getting production access](https://docs.publishing.service.gov.uk/manual/rules-for-getting-production-access.html) which requires engineers to do two 2ndline support shifts before they can be granted access.
+
+This can take a long time even in the best case, and doesn't work well in several common situations:
+
+- contractors tend not to do 2ndline shifts, as it's outside their scope of work
+- we make ad hoc access exceptions for senior hires (such as lead technical architects) who need some access to systems, but don't necessarily need full production admin access
+- experience of doing 2ndline shifts is not necessarily the best preparation for working with production systems, particularly for frontend developers
+- our process for when junior developers and apprentices should be given production access isn't clear
+
+There have been numerous cases where teams with only a few people with production access are slowed down because deployments become a bottleneck.
+
+[RFC 128 - Continuous Deployment](https://github.com/alphagov/govuk-rfcs/blob/main/rfc-128-continuous-deployment.md) has in some ways made the situation worse. We wanted to make sure that rolling out continuous deployment didn't inadvertently weaken a policy, so we disabled merge access for people without production access on repositories which were continuously deployed.
+
+There were [attempts at formalising this process around 2016](https://docs.google.com/document/d/1lo1JAkFIeWrIl-7bzkzps8rYi2hscV6IOpDzc8dhFLs/edit#). Unfortunately we didn't go into detail on why 2ndline shifts were required. The theory is that the logic went something like this:
+
+- In the early days of GOV.UK our deployment pipeline and monitoring was immature
+- This meant that debugging issues with deployments often required ssh access
+- In turn, this meant that only people with ssh access were allowed to do deployments
+- Because it's easy to make mistakes when using highly privileged access through ssh, we required people to do a couple of 2ndline shifts to make sure they knew what they were doing
+
+Note that by this logic, the 2ndline requirements are defending against accidental mistakes, not against malicious access. We have separate defences against malicious employees (security clearance, probation etc.).
+
+## Proposal
+
+We should implement a new level of access - "Production Deploy Access". This level of access should include:
+
+- Permission to deploy apps in Jenkins (ideally without Jenkins admin permission)
+- Permission to merge continuously deployed applications
+- Readonly access to AWS, Fastly, logging systems, etc.
+- admin access to GOV.UK Signon (create and edit normal users)
+
+This level of access should not include:
+
+- ssh permission to production or staging
```

Comment:
Should there also be a restriction on deploying Puppet?

Otherwise a deploy-only user could give themselves higher levels of access (e.g. SSH into prod machines) or make changes to Jenkins tasks by deploying a branch of puppet.

Reply:
>a deploy-only user could give themselves higher levels of access

This kind of depends on whether we're thinking of this as a guard against *malicious* users or simply careless/inexperienced ones.

And of course even with production deploy access they'd need someone else to review; it's not an absolute protection but should make it harder to make questionable changes.

Reply:
Yeah, this isn't intended to guard against malicious users - doing 2ndline shifts doesn't make you any more trustworthy.

There are plenty of ways a malicious user could use "deploy" access to do escalate to higher levels (I won't go in to them here because it's a public forum, but if it's worthwhile we could discuss them in a google doc addendum to this RFC).

The point about "they'd need someone else to review" is interesting - at the moment, the Deploy App job takes a commit-ish, so there's no protection (that I'm aware of) against someone asking Jenkins to deploy a branch (which doesn't have to be reviewed).

Reply:
We have on occasion wanted to deploy a branch to staging. I know this for sure but can't recall why. Production I'm with you, though.

Reply:
I'm going to resolve this thread without putting any special restrictions on puppet.

I do think there's a wider piece about tightening up our access controls, but it's very tied up with the replatforming project and the move to Argo. I'm sure we can improve on what we've got now.